### PR TITLE
Issue #10687 - WebSocket remembers mappings on restart

### LINKED ISSUE #10687: Jetty WebSocket remembers mappings on restart
**Jetty version(s)**
12+

**Jetty Environment**
`core`

**Description**
During `Server` start, a WebSocket endpoint is mapped to a pathspec.
Performing a `Server` restart (stop+start) results in an exception that the endpoint is already mapped.
This is because the mappings are not discarded or cleared upon restart.
The `eeN` environments do not have this problem.

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Fixes #10687

- Ensure the `WebSocketMappings` is cleared on restart of the server.
- Make the configurator on the `WebSocketUpgradeHandler` run in `doStart()` so configuration is reapplied when starting.

### CODE REVIEW

**Review by sbordet — CHANGES_REQUESTED** [review]
As discussed, let's add the `Consumer<ServerWebSocketContainer>` as a parameter to a new static method, like ee10.

File: jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java [hunk 1]
```diff
@@ -83,37 +83,79 @@ public class WebSocketUpgradeHandler extends Handler.Wrapper
...
+     * @deprecated use {@link #getContainer()} or {@link #from(Server, ContextHandler, Consumer)}.
      */
+    @Deprecated
     public WebSocketUpgradeHandler configure(Consumer<ServerWebSocketContainer> configurator)
     {
         configurator.accept(_container);
         return this;
     }
 
+    public ServerWebSocketContainer getContainer()
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java]: Rename to `getServerWebSocketContainer()`.

File: jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketUpgradeHandler.java [hunk 2]
```diff
@@ -24,17 +26,30 @@
...
-        this.mappings = new WebSocketMappings(components);
+        this(components, null);
+    }
+
+    public WebSocketUpgradeHandler(Consumer<WebSocketUpgradeHandler> configurator)
+    {
+        this(null, configurator);
+    }
+
+    public WebSocketUpgradeHandler(WebSocketComponents components, Consumer<WebSocketUpgradeHandler> configurator)
```
Comment (gregw) [inline jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketUpgradeHandler.java:48]: Need javadoc to clearly say when the configurator is called.

File: jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketUpgradeHandler.java [hunk 3]
```diff
@@ -24,17 +26,30 @@
 
 public class WebSocketUpgradeHandler extends Handler.Wrapper
 {
-    private final WebSocketMappings mappings;
-    private final Configuration.ConfigurationCustomizer customizer = new Configuration.ConfigurationCustomizer();
+    private final WebSocketMappings _mappings;
```
Comment (gregw) [inline jetty-core/jetty-websocket/jetty-websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketUpgradeHandler.java:29]: luv your underscore work!

File: jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java [hunk 4]
```diff
@@ -83,37 +83,79 @@ public class WebSocketUpgradeHandler extends Handler.Wrapper
...
+     * <p>Creates a new {@link WebSocketUpgradeHandler}.</p>
+     * <p>The {@link WebSocketUpgradeHandler} is not yet linked to the given
+     * {@link ContextHandler}, therefore the caller code must ensure that
+     * the returned {@link WebSocketUpgradeHandler} is a descendant of the
+     * given {@link ContextHandler}.</p>
+     *
+     * @param server the {@link Server} object used to lookup common WebSocket components
+     * @param context the {@link ContextHandler} ancestor of the returned {@link WebSocketUpgradeHandler}
+     * @param configurator a {@link Consumer} that is called to allow the {@link ServerWebSocketContainer} to
+     * be configured during the starting phase of the {@link WebSocketUpgradeHandler}.
```
Comment (gregw) [inline jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketUpgradeHandler.java:99]:
```suggestion
     * @param configurator a {@link Consumer} that is called to configure the {@link ServerWebSocketContainer}
     * during the starting phase of the {@link WebSocketUpgradeHandler}.   Mappings should be added to the container
     * in the confugurator on each start of the handler as they will not be preserved on any subsequent starts.
```

**Review by gregw — DISMISSED** [review]
Just some javadoc, otherwise good.

**Review by sbordet — APPROVED** [review]

**Pull request merged**